SOSS/dms/soss_centroids.py

```diff
@@ -0,0 +1,377 @@
+
+
+def get_uncontam_centroids_edgetrig(image, header=None, mask=None, poly_order=11,
+                                    halfwidth=5, mode='combined', verbose=False):
+    """Determine the x, y coordinates of the trace using the derivatives along the y-axis.
+    Works for either order if there is no contamination.
+
+    :param image: A 2D image of the detector.
+    :param header: The header from one of the SOSS reference files.
+    :param mask: A boolean array of the same shape as stack. Pixels corresponding to True values will be masked.
+    :param poly_order: Order of the polynomial to fit to the extracted trace positions.
+    :param halfwidth: the size of the window used when computing the derivatives.
+    :param mode: Which trace values to use. Can be 'maxedge', 'minedge', 'mean' or 'combined'.
+    :param verbose: If set True some diagnostic plots will be made.
+
+    :type image: array[float]
+    :type header: astropy.io.fits.Header
+    :type mask: array[bool]
+    :type poly_order: int
+    :type halfwidth: int
+    :type mode: str
+    :type verbose: bool
+
+    :returns: xtrace, ytrace, tracewidth, param - The x, y coordinates of trace as computed from the best fit polynomial
+    and the best-fit polynomial parameters.
+    :rtype: Tuple(array[float], array[float], array[float])
+    """
+
+    # If no mask was given use all pixels.
+    if mask is None:
+        mask = np.zeros_like(image, dtype='bool')
+
+    # Call the script that determines the dimensions of the stack.
+    dimx, dimy, xos, yos, xnative, ynative, padding, refpix_mask = \
+        get_image_dim(image, header=header, verbose=verbose)
+
+    # Replace masked pixel values with NaNs.
+    image_masked = np.where(mask | ~refpix_mask, np.nan, image)
+
+    # Use edge trigger to compute the edges and center of the trace.
+    ytrace_max, ytrace_min, ytrace_best, widths_best = edge_trigger(image_masked, halfwidth=halfwidth, yos=yos, verbose=verbose)
+
+    # Use different y-positions depending on the mode parameter.
+    if mode == 'maxedge':
+        ytrace = ytrace_max
+    elif mode == 'minedge':
+        ytrace = ytrace_min
+    elif mode == 'mean':
+        ytrace = (ytrace_min + ytrace_max)/2.
+    elif mode == 'combined':
+        ytrace = ytrace_best
+    else:
+        raise ValueError('Unknow mode: {}'.format(mode))
+
+    # Fit the y-positions with a polynomial and use the result as the true y-positions.
+    xtrace = np.arange(dimx)
+    mask = np.isfinite(ytrace)
+
+    # If no polynomial order was given return the raw measurements.
+    if poly_order is None:
+        param = []
+    else:
+        param = robust_polyfit(xtrace[mask], ytrace[mask], poly_order)
+        ytrace = np.polyval(param, xtrace)
+
+    # If verbose visualize the result.
+    if verbose is True:
+        _plot_centroid(image_masked, xtrace, ytrace)
+
+    return xtrace, ytrace, widths_best, param
+
+
+def test_uncontam_centroids_edgetrig():
+    """"""
```

Comment:
doc string + if temporary add TODO to remove

Reply:
Removed.